operators/src/util/statistics.rs

```
@@ -0,0 +1,723 @@
+    _phantom: PhantomData<T>,
+}
+
+impl<T> PSquareQuantileEstimator<T>
+where
+    T: AsPrimitive<f64>,
+{
+    /// Creates a new estimator for the given quantile. The estimator is
+    /// update with all valid values from the given `initial_samples`.
+    /// The initial samples must at least contain 5 valid elements
+    /// (i.e., 5 elements that do not translate to `f64::NAN`).
+    ///
+    /// # Panics
+    /// If the given quantile is not within the interval (0,1).
+    ///
+    /// # Errors
+    /// If the `initial_samples` contain less than 5 valid elements.
+    pub fn new(
+        quantile: f64,
+        initial_samples: &[T],
+    ) -> Result<PSquareQuantileEstimator<T>, StatisticsError> {
+        assert!(
+            quantile > 0.0 && quantile < 1.0,
+            "The desired quantile must be in the interval (0,1)"
+        );
+
+        // Initialize marker positions
+        let positions = [1, 2, 3, 4, 5];
+
+        // Initialize desired marker positions
+        let desired = [
+            1.0,
+            1.0 + 2.0 * quantile,
+            1.0 + 4.0 * quantile,
+            3.0 + 2.0 * quantile,
+            5.0,
+        ];
+        // Initialize marker increment
+        let increment = [0.0, quantile / 2.0, quantile, (1.0 + quantile) / 2.0, 1.0];
+
+        // Initialize marker values
+        let mut markers = [0.0; 5];
+        let mut iter = initial_samples.iter();
+        let mut sample_count = 0;
+
+        for v in &mut iter {
+            let v: f64 = v.as_();
+            if !v.is_nan() {
+                markers[sample_count] = v;
+                sample_count += 1;
+                if sample_count > 4 {
+                    break;
+                }
+            }
+        }
+
+        // We require at least 5 valid initial samples
+        if sample_count < 5 {
+            return Err(StatisticsError::Initialization {
+                reason: "Insufficient valid samples.".to_owned(),
+            });
+        }
```

Maybe check this early on and fail early.